src/inputfetcher.h

```diff
@@ -0,0 +1,197 @@
+                }
+                // Input missing or spent. This block will fail validation.
+                // Skip remaining inputs.
+                m_input_head.store(m_inputs.size(), std::memory_order_relaxed);
+                input.status.store(Input::Status::FAILED, std::memory_order_relaxed);
+            }
+            m_barrier.arrive_and_wait();
+        }
+    }
+
+public:
+    explicit InputFetcher(size_t worker_thread_count) noexcept
+        : m_barrier{static_cast<int32_t>(worker_thread_count + 1)}
+    {
+        for (size_t n{0}; n < worker_thread_count; ++n) {
+            m_worker_threads.emplace_back([this, n]() {
+                util::ThreadRename(strprintf("inputfetch.%i", n));
+                WorkLoop();
+            });
+        }
+    }
+
+    //! Fetch all block inputs from cache or db, and insert into temp_cache.
+    void FetchInputs(CCoinsViewCache& temp_cache, const CCoinsViewCache& cache, const CCoinsView& db, const CBlock& block) noexcept
+    {
+        if (block.vtx.size() <= 1 || m_worker_threads.size() == 0) {
+            return;
+        }
+
+        // Loop through the inputs of the block and set them in the queue.
+        // Construct the set of txids to filter, and count the outputs to reserve for temp_cache.
+        auto outputs_count{block.vtx[0]->vout.size()};
+        for (size_t i{1}; i < block.vtx.size(); ++i) {
+            const auto& tx{block.vtx[i]};
+            outputs_count += tx->vout.size();
+            m_txids.emplace(tx->GetHash());
+            for (const auto& input : tx->vin) {
+                m_inputs.emplace_back(input.prevout);
+            }
+        }
+
+        // Setup shared pointers and start workers.
+        m_db = &db;
+        m_cache = &cache;
+        m_input_head.store(0, std::memory_order_relaxed);
+        m_barrier.arrive_and_wait();
+
+        // Insert fetched coins into the temp_cache as they are set to READY.
+        temp_cache.Reserve(m_inputs.size() + outputs_count);
+        for (auto& input : m_inputs) {
+            auto status{input.status.load(std::memory_order_acquire)};
+            while (status == Input::Status::WAITING) {
+                std::this_thread::yield();
+                status = input.status.load(std::memory_order_acquire);
+            }
+            if (status == Input::Status::READY) {
+                temp_cache.EmplaceCoinInternalDANGER(COutPoint{input.outpoint}, std::move(input.coin));
+            } else if (status == Input::Status::FAILED) [[unlikely]] {
+                break;
+            }
+        }
+
+        m_barrier.arrive_and_wait();
+        // Cleanup after all worker threads have exited the inner loop.
+        m_txids.clear();
+        m_inputs.clear();
+        m_db = nullptr;
+        m_cache = nullptr;
+    }
+
+    ~InputFetcher()
+    {
+        m_request_stop = true;
+        m_barrier.arrive_and_wait();
```

Comment:
not certain, but I think this should likely be:
```suggestion
        m_barrier.arrive_and_drop();
```